### PR TITLE
Cling unlock depth

### PR DESCRIPTION
**Pull request by Axel-Naumann**
Do not reset the mutex to 0 readers, 0 writers, but to wherever the mutex was before entering a lock frame related to user code interpretation (which might in turn trigger a call to the interpreter).

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
@pcanal - happy now? ;-)

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -8453,3 +8446,65 @@ const char* TCling::TypedefInfo_Title(TypedefInfo_t* tinfo) const
...
+void TCling::ApplyToInterpreterMutex(void *delta)
+{
+   if (gInterpreterMutex) {
+      if (delta) {
+         auto typedDelta = static_cast<TVirtualRWMutex::StateDelta *>(delta);
+         std::unique_ptr<TVirtualRWMutex::StateDelta> uniqueP{typedDelta};
+         gCoreMutex->Apply(std::move(uniqueP));
+      }
+   }
+   fInitialMutex.pop_back();
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx]: Is the number of calls to ApplyToInterpreterMutex guaranteed by construction?  If not and ApplyToInterpreterMutex is called one too many time, the fInitialMutex could be empty and all the guard like
```
  if (fInitialMutex.back()) 
```
would have random behavior.

File: core/thread/src/TReentrantRWLock.cxx [hunk 2]
```diff
@@ -221,100 +221,158 @@ void TReentrantRWLock<MutexT, RecurseCountsT>::WriteUnLock(TVirtualRWMutex::Hint
...
-   const auto readerCount = pState->fReadersCount;
+   if (typedDelta->fDeltaWriteRecurse < 0) {
+      Error("TReentrantRWLock::Apply", "Negative write recurse count delta!");
+      return;
+   }
+   if (typedDelta->fDeltaReadersCount < 0) {
+      Error("TReentrantRWLock::Apply", "Negative read count delta!");
+      return;
+   }
+   AssertReadCountLocIsFromCurrentThread(typedDelta->fReadersCountLoc);
```
Comment (pcanal) [inline core/thread/src/TReentrantRWLock.cxx]: Consider having 'AssertReadCountLocIsFromCurrentThread' be (or wrapped into) a macro to make sure there is no (useless) function call when in release mode.

File: core/metacling/src/TCling.cxx (hunk 1 above)
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: By construction, but I added an `R__ASSERT`.

**Comment by Axel-Naumann** [comment]
Merged manually.

**Pull request merged**